### PR TITLE
Refactors (ApplicativeDo, record syntax, fetchWith)

### PR DESCRIPTION
**Pull request by Profpatsch**
These are a few code refactors, 

* use `ApplicativeDo` where possible
* use explicity record syntax instead of ordering-dependent constructors
* reduce complexity of `fetchWith` by moving most arguments into a single enum in multiple steps

### CODE REVIEW

File: src/Distribution/Nixpkgs/Fetch.hs [hunk 1]
```diff
@@ -211,11 +219,6 @@ fetchWith (supportsRev, kind, command, addArgs) source = do
                  Right ds -> return (Just (ds { derivKind = kind }, BS.unpack l))
  where
```
Comment (maralorn) [inline src/Distribution/Nixpkgs/Fetch.hs:270]: You accidentally produced an empty trailing `where`.

File: src/Distribution/Nixpkgs/Fetch.hs [hunk 2]
```diff
@@ -115,7 +117,7 @@ fromDerivationSource DerivationSource{..} =
 
 -- | Fetch a source, trying any of the various nix-prefetch-* scripts.
 fetch :: forall a.
-         Bool                                   -- ^ If True, fetch submodules when the source is a git repository
+         FetchSubmodules                                   -- ^ whether to fetch submodules when the source is a git repository
```
Comment (maralorn) [inline src/Distribution/Nixpkgs/Fetch.hs:120]: Looks like the comment is not aligned anymore.

**Review by maralorn — APPROVED** [review]
This PR is a delight. Thank you very much!

Just have two tiny nitpicks.

I’ll give others a chance to chime in as well. But from my point of view this is basically good to go.

Just out of curiosity: What was your motivation for this PR?

**Comment by sternenseemann** [comment]
(force-pushed to trigger CI with 8.8.4)

**Comment by Profpatsch** [comment]
> Just out of curiosity: What was your motivation for this PR?

I … accidentally looked at the code and had a bunch of time :sweat_smile:

**Comment by Profpatsch** [comment]
I had some weird rebasing issues, but now it should be rebased on master & the comments fixed

**Comment by maralorn** [comment]
Okay, thanks!

**Pull request merged**